### DISCUSSION TITLE
Best practice for optionally bundling a library dependency

### DISCUSSION THREAD

**Discussion by minrk** [mesonbuild/meson-python#556] [Q&A]
Split off from #410

### Summary

In general, I'd like to get to a documented best practice for _optionally_ bundling a library dependency. At a high level:

```python
if library.found():
    link_it()
else:
    build_and_bundle_it()
```

In #410, [@rgommers mentioned](https://github.com/mesonbuild/meson-python/discussions/410#discussioncomment-8011565) static linking for the bundle branch, which simplifies the bundling issue, since meson doesn't need to duplicate the delvewheel/delocate step. I _believe_ this should work for me, but figuring out the Windows version needs some work

### Specific case

I'm exploring updating the build system for pyzmq, which bundles libzmq _if not found_. This is slightly more complex than #410 because libzmq itself has a dependency on libsodium and both should be bundled. It is _further_ complicated by the need to support Windows, and the fact that libsodium has [visual studio solutions](https://github.com/jedisct1/libsodium/tree/1.0.19/builds/msvc) on Windows while ZeroMQ uses [CMake](https://github.com/zeromq/libzmq/tree/v4.3.5/builds/cmake). So far, it seems like it should be roughly:

```python
if libzmq_found:
    locate libs, includes with pkg-config/etc. (standard, simple)
else:
    fetch libsodium, libzmq
    if Windows:
        build static libsodium with msbuild
        build static libzmq against static libsodium with cmake
    else:
        build static libsodium with autotools
        build static libzmq against static libsodium with autotools

if PyPy:
    _zmq = cffi extension
else:
    _zmq = cython extension

if libzmq_found:
    dynamically_link(_zmq, libzmq)
else:
    statically_link(_zmq, local libzmq)
```

### more info

https://github.com/oscarbenjamin/mesontest provides a useful example for using wrapfiles to set up the fallback builds. It's gotten me pretty far for the simple posix case, at least.

**Comment by rgommers** [comment]
> I _believe_ this should work for me, but figuring out the Windows version needs some work

One of the thoughts I had, also because your build dependencies have different build system usage on Windows, is whether you actually want/need to support this on Windows? In general you need it because of folks building from source when running `pip install pyzmq` (e.g., because there are no wheels on PyPI). However, if you do have wheels on PyPI, it may be less work to simply document that Windows users who want to build from source need to prepare `libzmq`/`libsodium` in the same way as in your wheel build env.

There is anyway a second Windows issue: if you do `pip install .` _locally_ it is entirely possible for `libzmq` to be found, linked against, and then go missing at runtime (this is not specific to Meson). The reason is that Windows doesn't have RPATH support, so `pip` is basically entirely broken for this use case - it will not work until you add manual handling of the external dependency or run `delvewheel` over the produced wheel. We have this exact problem with OpenBLAS in numpy/scipy, and always have some weird workarounds which involve copying the shared library around manually and adding the directory in which we put it to the DLL search path.

> and the fact that libsodium has [visual studio solutions](https://github.com/jedisct1/libsodium/tree/1.0.19/builds/msvc) on Windows while ZeroMQ uses [CMake](https://github.com/zeromq/libzmq/tree/v4.3.5/builds/cmake).

Other build systems: the most robust option is having upstream support for building with Meson, or a [WrapDB](https://mesonbuild.com/Wrapdb-projects.html) file to do so. However, Meson also [understands CMake](https://mesonbuild.com/CMake-module.html#cmake-subprojects) - unless the CMake files from the upstream repo do really exotic things, Meson can parse those upstream `CMakelists.txt` files and auto-translate them to Meson on the fly when building.

Other build systems are harder, I believe that cannot be done as part of a subproject build. So `libzmq` with autotools may be an issue. However, there is a WrapDB file for `cppzmq` already, which may be of use.

> `if PyPy:`

This is the second time in two days I've come across the need for determining whether the interpreter is PyPy. The one-liner from https://github.com/numpy/numpy/pull/25531 to implement this is:
```meson
if not py.get_variable('SOABI').startswith('pypy')
```
but it looks like it would be nice to add a `py.interpreter_name()` method to Meson (@eli-schwartz WDYT?).

> if `dynamically_link` else `statically_link`

There is no need for the explicit if-else at this point. You'll simply use a `libzmq_dep` dependency object, and Meson already knows whether that is static or dynamic and will link it correctly.

Instead, I think you do it in the discovery phase, something like:
```meson
libzmq_dep = dependency('zmq',
    fallback : ['zmq', 'libzmq_dep'],
    default_options: ['default_library=static']
)
```
as documented at https://mesonbuild.com/Subprojects.html#toggling-between-system-libraries-and-embedded-sources.

> https://github.com/oscarbenjamin/mesontest provides a useful example for using wrapfiles to set up the fallback builds. It's gotten me pretty far for the simple posix case, at least.

We should indeed add something like this to the docs, going into a bit more detail about what is specific to distributing wheels on PyPI (since that's the hard part and not covered by the Meson docs).

**Reply by eli-schwartz** [reply]
> but it looks like it would be nice to add a `py.interpreter_name()` method to Meson (@eli-schwartz WDYT?).

I have no real objections, but "interpreter" name is kind of vague and it should probably be "implementation" instead.

**Reply by minrk** [reply]
Lacking other reasons, it would probably be nice to match Python's own naming and values for this, which would suggest `implementation` and use the value from [platform.python_implementation()](https://docs.python.org/3/library/platform.html#platform.python_implementation).

**Reply by rgommers** [reply]
Sounds good. Moving that to https://github.com/mesonbuild/meson/issues/12727.

**Comment by minrk** [comment]
Thanks for the tips. Unfortunately, I do indeed need to support this on Windows. That is in fact the _most_ important case.

I've gotten pretty far. Unfortunately, somethings that's causing me what seems like an unreasonable amount of trouble is getting one ExternalProject to depend on another one. ExternalProjects can depend on Targets, but can't depend on a Dependency, and can only _get_ a `Dependency` from the ExternalProject object (the `.target` public attribute is not exposed), and I don't know enough meson to turn a Dependency back into a Target so it can be depended upon. All I want to do is make one subproject build before the other, but I can't seem to figure it out.

**Reply by eli-schwartz** [reply]
This is the downside of Meson's "external projects" support, because meson doesn't natively understand the external project's configure stage. It is (currently?) unable to inject one dependency into another. It's not a simple matter of building one before the other, either. Without a way to coordinate knowledge about targets that will be built in the future, there's no good way for the external project to look up a dependency.

I'd be interested to see the issue solved but I have no idea how one would attempt to model the problem.

**Reply by minrk** [reply]
This is probably not a good idea, but I went with building an [empty static library](https://github.com/zeromq/pyzmq/pull/1922/files#diff-44b81d1b177a1b7074f7c3f892b3c0aa65bcef673d6c89150b4c208e73036cdd) that [depends on the actual library](https://github.com/zeromq/pyzmq/pull/1922/files#diff-6dceec76083dd85d64c414a600e3e25b3b91a4cb31bd32fc4ed4813d47af5157R27), and then [depending on that](https://github.com/zeromq/pyzmq/pull/1922/files#diff-4e3db3fe66198993c4216809e77b2fcabee562362d42b393e936cf74af99a6adR25). This gave me a target that depends on the dependency. Fortunately, I didn't need to actually extract cflags, libs, etc. from the dependency as inputs to the second configure, since I also can't seem to figure out how to do that, which seems like it should be a simple attribute or method on the dependency object.

At least for now, this seems to be good enough. It would have been much simpler if the [ExternalProject.target public attribute](https://github.com/mesonbuild/meson/blob/e4bbc630b67ef97ad842badd00855e64cff12e13/mesonbuild/modules/external_project.py#L229) were accessible in meson, but it doesn't seem to be. That's all I really needed.

Since I need to have custom targets for the msbuild and (maybe not?) cmake builds on Windows, perhaps custom targets for `./configure; make; make install` will work better than ExternalProject. I'm not sure.

**Comment by minrk** [comment]
https://github.com/zeromq/pyzmq/pull/1922 has working builds, at least for the simple cases of Cython on linux/mac

**Reply by minrk** [reply]
Are there any examples of building CFFI extensions with meson?

**Reply by eli-schwartz** [reply]
https://pypi.org/project/pypkgconf/

**Reply by eli-schwartz** [reply]
Why is GitHub offering to translate that link from Portuguese to English.

**Reply by minrk** [reply]
Perfect, thanks! Looks like trading the `compile` in my existing script for `emit_c_code` is ~all I need to do.

**Reply by mattip** [reply]
A bit off-topic, but maybe adding a sample meson.build to the cffi documentation would be nice?

**Reply by kivkiv12345** [reply]
Using the linked meson.build as inspiration, I manged to bundle a library object with custom_target():
```
custom_target('<bundled_target_name>',
		output: '<bundled_target_file>.so',
		install: true,
		install_dir: py.get_install_dir() / '<bundled_target__init__py_subdir>',
		input: <bundled_lib>,
		command: ['cp', '@INPUT@', '@OUTPUT@'],
		build_by_default: true
	)
```
`py.get_install_dir()` fixes the `meson-python: error: Could not map installation path to an equivalent wheel directory: '{prefix}'` error. Funnily enough it points to `/usr/local/lib/python3/dist-packages/`, but the .whl is installed to `~/.local/lib/python3.12/site-packages/` which is quite neat.

**Comment by minrk** [comment]
Any idea how to prevent things from running on sdists? Meson is running various library detection, etc. steps during sdists, when of course it shouldn't be looking for any dependencies just to build a tarball. But I can't find any documentation about how to specify things to be run or not run during different stages, or even how to detect what stage is being run.

**Reply by eli-schwartz** [reply]
Meson's builtin dist command runs as a custom target on the same build directory that you use to compile and install. It needs to:
- because dist can depend on which dependencies were found (and which were acquired via subprojects instead)
- to parse the version number
- to register and execute `meson.add_dist_script()`

IIRC meson-python uses a throwaway build directory for this?

**Reply by rgommers** [reply]
It is actually slightly annoying indeed that the configure stage runs. When you have complex dependencies and things like compiler flag testing, it makes it both fragile and slow to create an sdist. 

From a naive point of view, just getting straight to all the git-tracked files and stuffing those into a tarball would be nicer (this is what `scikit-build-core` and `hatchling` do). However, it's pretty hard to see `meson-python` making that transition at this point. The first two bullets in the comment above can be done in another way with a bit of effort, but `add_dist_script` is harder to replace; at that point we'd have to come up with a custom plugin/scripting type thing.

In the end it's not too big of an issue I'd say. Slow is relative (`numpy`'s configure stage is ~10 seconds, which is probably about as slow as it gets), so the main issue is robustness. It adds a requirement that creating an sdist needs not just `meson-python`, but all required dependencies of the project. It's pretty rare that one doesn't have such an environment at hand though, because it's needed anyway to build wheels.

**Reply by dnicolodi** [reply]
> of course it shouldn't be looking for any dependencies just to build a tarball

As @eli-schwartz explained, this is of course not the case. meson-python relies on `meson dist` to create the sdist archive and `meson dist` requires to configure the project.

Why do you want to avoid running `meson setup` during the sdist build?

> IIRC meson-python uses a throwaway build directory for this?

That's correct: meson-python uses a throwaway build directory unless one is specified. If a build directory is specified with the `build-dir` option, the project is reconfigured, but Meson in this case should be able to reuse cached results from previous `meson setup` runs.

**Reply by dnicolodi** [reply]
> The first two bullets in the comment above can be done in another way with a bit of effort

One of the things that meson-python optionally do is to take the version number from the Meson project metadata. This is available is an easy to access way (in a Meson introspection file) only after `meson setup`. As the version in a Meson project may be dynamic, implementing this outside of Meson would require reimplementing quite a bit of functionality.

**Reply by eli-schwartz** [reply]
The larger issue here to me, personally, is that I think it is strange to want to build an sdist unless you are the developer and are seeking to make a release. This is uncommon enough that the overhead isn't a big deal to me, personally.

I'll say that I think it was a mistake for PyPA to standardize on an interface for making them, because it encourages people to make sdists when they shouldn't (looking at you, `build`...).

Meson partially circumvents this by not *letting* you build an sdist from another sdist. If you try, it's an error because `meson dist` requires a VCS repository.

**Reply by dnicolodi** [reply]
@eli-schwartz My understanding is that `build` by default builds wheels from sdists to make sure that the sdist is complete. `build` was thought to be the tool used by developers to build sdists and wheels to upload to PyPI, thus this sanity check makes sense, especially if your main focus are pure Python packages that take a fraction of a second to build.

My main issue with the Python packaging ecosystem is the assumption that source distributions and binary distributions are interchangeable and that you never know how a package will be installed before running `pip install` AFAIK, there are `pip` options to disable using binary distributions, but not to disable using source distributions.

**Reply by rgommers** [reply]
>  but not to disable using source distributions.

There's `pip install --only-binary`.

> My main issue with the Python packaging ecosystem is the assumption that source distributions and binary distributions are interchangeable

100% agreed

**Reply by minrk** [reply]
okay, i understand that it runs the configure stage. In that case, how can I _detect_ that the configure stage is running for a dist and not a build, so I can give it more lenient dependency rules. Because for my project, dist definitely should _not_ be sensitive to any dependencies. This only came up because it's _failing_ to run stages that are not relevant to the dist (libzmq is unavailable, so it proceeds to attempt to bundle dependencies via ExternalProject, which in turn fails to run a command whose output will be unused). I care a lot less about time, but if I can tell meson to skip looking for dependencies when building an sdist, that would probably be enough. But I can't seem to figure out what I can tell meson and/or what meson can tell me to distinguish the two runs.

**Reply by rgommers** [reply]
>  if I can tell meson to skip looking for dependencies when building an sdist

I don't think there's a way of doing that. Given how things are layered, that also cannot be possible: first `meson setup` is run, and then `meson dist -Cbuild-dir` follows. Dependency detection happens in the `setup` part, which cannot know if `dist` or `compile` comes next.

 > This only came up because it's failing to run stages that are not relevant to the dist (libzmq is unavailable, so it proceeds to attempt to bundle dependencies

That sounds like a problem indeed. So the question is how to tell Meson to include only the contents of `subprojects/` into your sdist without changes, but instead it vendors what it downloaded during the `setup` stage? I don't have much experience with this, but I'd hope one of the `--wrap-mode` values does the right thing here: https://mesonbuild.com/Creating-releases.html#include-subprojects-in-your-release.

**Reply by dnicolodi** [reply]
AFAIK, this is the default behavior of Meson. If you want to include the subprojects source code, you need to specify the `--include-subprojects` option to `meson dist`.

**Reply by minrk** [reply]
Thanks for the help! I specifically _don't_ want to include subprojects in the dist, so that's not what I'm after. I can see how dependency detection should run before dist if that were the case.

 It seems like meson maybe isn't the tool for me, since its abstractions and strong opinions are fighting me at every turn.

**Reply by dnicolodi** [reply]
@minrk Yuo are not doing a very good job in explaining what you are trying to do, how you are trying to do it, and what does not work as expected. Therefore it is extremely difficult to help you. Simply stating that the solution you have tried does not work for solving your problem does not help without understanding what problem you are trying to solve.: you may be approaching the problem in the wrong way.

In the specific case of building an sdist, I don't understand where your problem is coming from. Apparently you are trying to build an sdist on a system where `meson setup` for your project fails. The first question is: why are you trying to build sdists on such a system? I don't see any reason to do so. The second question is: why is `meson setup` failing?

**Reply by eli-schwartz** [reply]
> but if I can tell meson to skip looking for dependencies when building an sdist, that would probably be enough. But I can't seem to figure out what I can tell meson and/or what meson can tell me to distinguish the two runs.

The simple and straightforward way to do that would be to add a new meson option `-Dsdist-stub=true` that includes whatever meson.build content you need to make a sdist correctly (this *might* just be "project()" if you don't have any dist scripts) and then uses meson's `subdir_done()` function inside a conditional -- and pass that argument to `-Csetup-args=....` when building an sdist only.

This still invites the question of what system are you trying to build an sdist on, if you specifically don't want to build a wheel on that system, and "maybe you should just not do that?" -- I do understand that that question is a personal opinion.

**Reply by eli-schwartz** [reply]
Also, to be clear -- meson is not strongly opinionated on how you should distribute your source tarballs, but includes a meson equivalent to the traditional autotools `./configure && make distcheck` and this includes doing a full test run of the created tarball by default.

There are people who use meson and don't use `meson dist` :) but since python requires a dist tarball for PyPI and also requires the build backend information that isn't easily available otherwise be baked into the tarball, this is more complex than it has to be.

In fact, I'd argue that python is the opinionated thing here. *Python* is the thing saying "there is only one way to release sdists and that is to go through the full power of a build backend including the calculation of dynamic metadata, and expose that to all users as the first step of compiling the code".

**Comment by minrk** [comment]
> There are people who use meson and don't use meson dist

Indeed, looks like I'd like to be one of those people! Maybe that's a feature request for meson-python - to do a more Python-friendly sdist instead of `meson dist`. I'm guessing I'm in the very small minority where it makes a difference, though, and a clear enough workaround is fine for me.

> pass that argument to -Csetup-args=.... when building an sdist only.

I think that's what I was looking for, thank you, as long as I can put this in pyproject.toml so meson-python always passes it during sdists. That's firmly on meson-python's side of things, rather than meson itself. The important thing is that it isn't passed on the command-line by me, it's specified in pyproject.toml, so it happens consistently.

scikit-build-core has [SKBUILD_STATE](https://scikit-build-core.readthedocs.io/en/latest/cmakelists.html#accessing-information), to tell the build files a bit about what's going on, if they need to act on it. A similar signal from meson-python seems quite sensible and practical to me.

> Yuo are not doing a very good job 

Nice. I [did my best](https://github.com/mesonbuild/meson-python/discussions/556#discussion-6031166) to describe the use case and [linked to the code itself and the builds that are failing](https://github.com/mesonbuild/meson-python/discussions/556#discussioncomment-8022668), but to reiterate, I am trying to:

1. build two dependencies with autotools (optionally, as fallback)
2. one of those must not run configure before build completes for the other because it is a dependency (pyzmq -> libzmq -> libsodium)
3. neither should run configure during sdist (both because it's irrelevant, and because one has to be built before the other can run configure)

Right now, sdist is [failing](https://github.com/zeromq/pyzmq/actions/runs/7897965246/job/21554447879?pr=1922#step:5:491) because ExternalProject is running configure for the _second_ autotools project as part of setup without having built the first. Neither autotools project will be in the dist, so all of this is irrelevant anyway, which is why I want to skip it.

I think perhaps a purely custom target could do the job where ExternalProject doesn't seem to have the necessary controls.

**Reply by rgommers** [reply]
> Indeed, looks like I'd like to be one of those people! Maybe that's a feature request for meson-python - to do a more Python-friendly sdist instead of `meson dist`.

I've heard this request a couple of times now, so I'll see if I can come up with a good problem summary in a new issue with an enhancement request. There are certainly trade-offs here that are being made. We get functionality for free from Meson, and dist scripts in particular can solve things that required hacky `build_ext` subclasses with distutils/setuptools. But on the other hand, the way things work now can be slower and at least surprising to folks coming from other Python build backends.

> one of those must not run configure before build completes for the other because it is a dependency (pyzmq -> libzmq -> libsodium)

Just checking: this does work fine when you build a wheel? So it is possible to do configure + build for subproject 1 before configuring subproject 2? I imagine so because it's documented at https://mesonbuild.com/Subprojects.html#subprojects-depending-on-other-subprojects, but if that case prevents `meson dist` from working as intended, maybe that should be documented.

**Reply by minrk** [reply]
> Just checking: this does work fine when you build a wheel? 

It had, but possibly not in the failing environment, because when it succeeded, it is possible the libzmq build may have found already-installed libsodium elsewhere on my system, in which case it wouldn't fail, but would produce an incorrect result.

Ultimately, I've [gone with cmake via scikit-build-core](https://github.com/zeromq/pyzmq/pull/1927), which has most of the same challenges, but it happened to be easier for me to muddle my way toward what i needed, which turned out to be:

1. actually build static libsodium during the 'setup' step. This violates the intended separation of setup and build, but works out in practice because the  'setup' phase is exclusively run immediately prior to actual wheel builds and has all the same info available to do it, and is required because the 'setup' of libzmq requires libsodium to be built already
2. import libzmq-static cmake targets using FetchContent _after_ libsodium is installed

Now that I've done it with cmake, I imagine the same approach ought to work in meson. I'd just need to specify the stub setup for sdist from pyproject.toml, and to understand how to import cmake into meson, which [seems straightforward](https://mesonbuild.com/CMake-module.html).

**Reply by eli-schwartz** [reply]
How does the cmake setup handle using a system libzmq (mandatory for redistributors that ship their own dependency management instead of using PyPI wheels)?

> actually build static libsodium during the 'setup' step. This violates the intended separation of setup and build, but works out in practice because the 'setup' phase is exclusively run immediately prior to actual wheel builds and has all the same info available to do it, and is required because the 'setup' of libzmq requires libsodium to be built already

This is also a bit awkward as it doesn't parallelize well -- and parallelism is one of the biggest advantages of moving to a build system that groks C/C++ at a more sophisticated level than the very basics as handled by setuptools.

**Reply by minrk** [reply]
> How does the cmake setup handle using a system libzmq (mandatory for redistributors that ship their own dependency management instead of using PyPI wheels)?

Basically the same as `dependency('libzmq')` and `cc.find_library('libzmq')` in the meson build. It looks for cmake config files, pkg-config, or finally a shared library on a path, and links with what it finds. This is the most common case.

The static build is only the 'fallback', used when no suitable system libzmq is found.
